### DISCUSSION TITLE
CustomType not working in MySQL

### DISCUSSION THREAD

**Discussion by razajamil** [mikro-orm/mikro-orm#2091] [Q&A] [answered]
**Describe the bug**
I'm trying to use a custom type for using ULID as PK stored as binary in MySQL. The custom type does not seem to have any effect at all, no logging, no error.

**Stack trace**

**To Reproduce**
```
// custom type in question
export class ULIDType extends Type<string, Buffer> {
  convertToDatabaseValue(value: string): Buffer {
    console.log('a')
    try {
      const buffer = idToBin(value)
      return buffer
    } catch (error) {
      throw ValidationError.invalidType(ULIDType, value, 'JS')
    }
  }

  convertToJSValue(value: Buffer): string {
    console.log('b')
    try {
      const id = binToId(value)
      return id
    } catch (error) {
      throw ValidationError.invalidType(ULIDType, value, 'database')
    }
  }

  getColumnType() {
    console.log('c')
    return 'binary(16)'
  }
}

// entity that uses ULIDType for PK
@Entity({ tableName: 'ulid_test' })
export class TestEntity {

  // this column is stored as binary(16) in table
  @PrimaryKey({
    type: ULIDType,
  })
  id: string

  @Property()
  name: string
}

// a simple query 
const results = await orm.em.find(TestEntity, {});
```

**Expected behavior**
TestEntity query should return objects with id as string.

**Actual behavior**
TestEntity query returns objects with id as Buffer.
```
[
  TestEntity {
    id: <Buffer 01 7a fe ca fb 65 65 cc 5e 6c b2 56 e4 c5 4a 93>,
    name: 'john'
  }
]
```

**Additional context**
The idToBin(), binToId() and id() functions work fine. If I ignore the typescript warnings and jam the parameters I'm able to insert and query correctly. 
The fact that there is logs aren't firing means I might be missing something basic but I've tried many different examples from scratch and it just doesn't work.

**Versions**

| Dependency | Version |
| - | - |
| node | 12.18.3 |
| typescript | 4.3.5 |
| mikro-orm | 4.5.7 |
| your-driver | MySQL |

**Comment by B4nan** [comment, marked as answer]
If you dont see any logs, it means the type was not considered as custom type. That can happen if there are multiple `Type` classes installed in your project, as the check is done via `instanceof` and it needs to be the same class. 

In general this is caused by wrong dependency resolution, e.g. you use `mikro-orm` instead of `@mikro-orm/core`, or when having misaligned package versions (e.g. `@mikro-orm/core` on 4.5.6 and `@mikro-orm/mysql` on 4.5.7).

Ensure you have only one installation of `@mikro-orm/core` in your node_modules.

**Reply by razajamil** [reply]
I see, I will check based on this. Many thanks.

**Reply by razajamil** [reply]
> Ensure you have only one installation of `@mikro-orm/core` in your node_modules.

This was the issue due to it being a monorepo. The solution was to create a separate @my-projects/mikro-orm and export all core/mysql and any other mikro-orm packages and reference everything from the same package.

Thank you Martin for all your hard work 🙏

**Reply by B4nan** [reply]
I'll try to get around this in v5, same issues were present with `Collection` and some other classes that were checked via `instanceof`, we can just add some symbol property and check that instead. Those issues are nightmare, very hard to catch without seeing the full project.
